### PR TITLE
[AIRFLOW-2099] Handle getsource() calls gracefully

### PR DESCRIPTION
**Pull request by night0wl**
There are several scenarios where Task Instance view tries to render
Python callables where 'x' is not the correct artefact to target.

This commit adds a helper fuction to test for known scenarios, and
derives the source from the correc artefact or as a default returns
'No source available for <type>'. This means that even in unknown or
unfixable edge cases, the Task Instance view still renders instead of
displaying an exception.

Make sure you have checked _all_ steps below.

### JIRA
- [x] My PR addresses the following [Airflow JIRA](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "[AIRFLOW-XXX] My Airflow PR"
    - https://issues.apache.org/jira/browse/AIRFLOW-XXX

### Description
- [x] Here are some details about my PR, including screenshots of any UI changes:
1. Adds a wrapper function for getting Python source code.
2. It tests for/catches known anomalies and implements a work around, finally returning 'source code not available for <type>' if no option works.

### Tests
- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
1.  test_dag_view_task_with_python_operator_using_partial
2. test_dag_view_task_with_python_operator_using_instance

### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

- [x] Passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

### CODE REVIEW

File: tests/core.py [hunk 1]
```diff
@@ -1800,6 +1800,20 @@ def test_fetch_task_instance(self):
         response = self.app.get(url)
         self.assertIn("run_this_last", response.data.decode('utf-8'))
 
+    def test_dag_view_task_with_python_operator_using_partial(self):
+        response = self.app.get(
+            '/admin/airflow/task?'
+            'task_id=test_dagrun_functool_partial&dag_id=test_issue_2099_dag&'
+            'execution_date={}'.format(DEFAULT_DATE_DS))
+        self.assertEquals(response.status_code, 200)
```
Comment (ashb) [inline tests/core.py]: Some kind of tests of the content would be good.

**Comment by night0wl** [comment]
Fair comment @ashb , I have amended the tests to use `AssertIn` instead, and to check for doc string of expected source code.

File: tests/dags/test_issue_2099.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
+# -*- coding: utf-8 -*-
```
Comment (bolkedebruin) [inline tests/dags/test_issue_2099.py:1]: Please use a descriptive name for the file and dag. This makes it easier to make out when it fails.

File: tests/dags/test_issue_2099.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    """ A function with two args """
+    pass
+
+
+partial_function = functools.partial(a_function, arg_x=1)
+class_instance = CallableClass()
+
+logging.info('class_instance type: {}'.format(type(class_instance)))
+
+dag = DAG(dag_id='test_issue_2099_dag', default_args=default_args)
```
Comment (bolkedebruin) [inline tests/dags/test_issue_2099.py:49]: Please use descriptive naming

File: airflow/www/views.py [hunk 4]
```diff
@@ -240,6 +240,31 @@ def wrapped_markdown(s):
     return '<div class="rich_doc">' + markdown.markdown(s) + "</div>"
 
 
+def get_python_source(x):
```
Comment (bolkedebruin) [inline airflow/www/views.py:243]: please move this to www/utils.py .

**Review by bolkedebruin — CHANGES_REQUESTED** [review]
Couple of small things, almost ready.

**Comment by bolkedebruin** [comment]
@night0wl any updates?

**Comment by bolkedebruin** [comment]
@night0wl Any updates? will close in next 3 days otherwise

**Comment by r39132** [comment]
@night0wl Kindly reopen when ready for review.

**Comment by night0wl** [comment]
Hi @r39132 and @bolkedebruin 

Apologies it has taken me ages to get back onto this.

I have rebased onto master branch, actioned the requests above and opened a new PR
https://github.com/apache/incubator-airflow/pull/3571

All Travis checks are passing.
